components/raftstore/src/store/fsm/peer.rs

```diff
@@ -898,6 +900,71 @@ where
             .unsafe_recovery_maybe_finish_wait_apply(/* force= */ self.fsm.stopped);
     }
 
+    // func be invoked firstly after assigned leader by BR, wait all leader apply to
+    // last log index func be invoked secondly wait follower apply to last
+    // index, however the second call is broadcast, it may improve in future
+    fn on_snapshot_recovery_wait_apply(&mut self, syncer: SnapshotRecoveryWaitApplySyncer) {
+        if self.fsm.peer.snapshot_recovery_state.is_some() {
+            warn!(
+                "can't wait apply, another recovery in progress";
+                "region_id" => self.region_id(),
+                "peer_id" => self.fsm.peer_id(),
+            );
+            syncer.abort();
+            return;
+        }
+
+        let target_index = self.fsm.peer.raft_group.raft.raft_log.last_index();
+
+        info!(
+            "snapshot recovery started";
+            "region_id" => self.region_id(),
+            "peer_id" => self.fsm.peer_id(),
+            "target_index" => target_index,
+            "applied_index" => self.fsm.peer.raft_group.raft.raft_log.applied,
+
+        );
+
+        // during the snapshot recovery, broadcast waitapply, some peer may stale
+        if !self.fsm.peer.is_leader() {
+            info!(
+                "snapshot follower recovery started";
+                "region_id" => self.region_id(),
+                "peer_id" => self.fsm.peer_id(),
+                "target_index" => target_index,
+                "applied_index" => self.fsm.peer.raft_group.raft.raft_log.applied,
+                "pending_remove" => self.fsm.peer.pending_remove,
+                "voter" => self.fsm.peer.raft_group.raft.vote,
+            );
+
+            // do some sanity check
+            // if it is learner during backup and never vote before, vote is 0
+            // if peer is suppose to remove
+            if self.fsm.peer.raft_group.raft.vote == 0 || self.fsm.peer.pending_remove {
+                info!(
```

Comment:
I think this case is possible:
The follower has voted successfully, but hasn't receive the new leader's first entry. So at the time there could be a gap (maybe large) between the leader and the follower.

To resolve the problem, we can add a condition: `peer.raft_group.raft.term == peer.raft_group.raft.last_log_term`.

Reply:
here we handle some corner cases, it looks we should handle peer.raft_group.raft.term != peer.raft_group.raft.last_log_term? leader applied to the last log, if the follower can't receive MsgAppend, it's term != last log term, we skip it?

Reply:
what about peer.raft_group.raft.term > peer.raft_group.raft.last_log_term? since the same term, the follower needs to apply to the last log during the term.

Reply:
proposal  peer.raft_group.raft.term > peer.raft_group.raft.last_log_term, and also add more comments in code